src/xdp/ebpfextension.c

```diff
@@ -0,0 +1,277 @@
+    )
+{
+    ((EBPF_EXTENSION_CLIENT *)Client)->ProviderData = (VOID *)ProviderData;
+}
+
+VOID *
+EbpfExtensionClientGetProviderData(
+    _In_ const EBPF_EXTENSION_CLIENT *Client
+    )
+{
+    return Client->ProviderData;
+}
+
+const ebpf_extension_dispatch_table_t *
+EbpfExtensionClientGetDispatch(
+    _In_ const EBPF_EXTENSION_CLIENT *Client
+    )
+{
+    return Client->ClientDispatch;
+}
+
+const VOID *
+EbpfExtensionClientGetClientContext(
+    _In_ const EBPF_EXTENSION_CLIENT *Client
+    )
+{
+    return Client->ClientBindingContext;
+}
+
+const ebpf_extension_data_t *
+EbpfExtensionClientGetClientData(
+    _In_ const EBPF_EXTENSION_CLIENT *Client
+    )
+{
+    return Client->ClientData;
+}
+
+VOID
+EbpfExtensionDetachClientCompletion(
+    _In_ EBPF_EXTENSION_CLIENT *Client
+    )
+{
+    TraceEnter(TRACE_CORE, "Client=%p", Client);
+
+    ASSERT(Client != NULL);
+    _Analysis_assume_(Client != NULL);
+
+    NmrProviderDetachClientComplete(Client->NmrBindingHandle);
+
+    TraceExitSuccess(TRACE_CORE);
+}
+
+static
+NTSTATUS
+EbpfExtensionProviderAttachClient(
+    _In_ HANDLE NmrBindingHandle,
+    _In_ const VOID *ProviderContext,
+    _In_ const NPI_REGISTRATION_INSTANCE* ClientRegistrationInstance,
+    _In_ const VOID *ClientBindingContext,
+    _In_ const VOID *ClientNpiDispatch,
+    _Outptr_ VOID **ProviderBindingContext,
+    _Outptr_result_maybenull_ const VOID **ProviderDispatch
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+    EBPF_EXTENSION_PROVIDER *Provider = (EBPF_EXTENSION_PROVIDER *)ProviderContext;
+    EBPF_EXTENSION_CLIENT *Client = NULL;
+    const ebpf_extension_dispatch_table_t *ClientDispatch;
+
+    TraceEnter(TRACE_CORE, "ProviderContext=%p", ProviderContext);
+
+    if ((ProviderBindingContext == NULL) || (ProviderDispatch == NULL) || (Provider == NULL)) {
+        Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    *ProviderBindingContext = NULL;
+    *ProviderDispatch = NULL;
+
+    Client = ExAllocatePoolZero(NonPagedPoolNx, sizeof(*Client), XDP_POOLTAG_EBPF_NMR);
+    if (Client == NULL) {
+        Status = STATUS_NO_MEMORY;
+        goto Exit;
+    }
+
+    Client->NmrBindingHandle = NmrBindingHandle;
+    Client->ClientModuleId = ClientRegistrationInstance->ModuleId->Guid;
+    Client->ClientBindingContext = ClientBindingContext;
+    Client->ClientData =
+        (const ebpf_extension_data_t *)ClientRegistrationInstance->NpiSpecificCharacteristics;
+
+    ClientDispatch = (const ebpf_extension_dispatch_table_t *)ClientNpiDispatch;
+    if (ClientDispatch == NULL) {
+        Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    Client->ClientDispatch = ClientDispatch;
+    Client->ProviderContext = Provider;
+
+    if (Provider->AttachCallback != NULL) {
+        Status = Provider->AttachCallback(Client, Provider);
+    }
+
+    if (NT_SUCCESS(Status)) {
+        Status = STATUS_SUCCESS;
+    } else {
+        Status = STATUS_ACCESS_DENIED;
```

Comment:
I'm curious why you do this? Perhaps a small comment explaining why you overwrite the attach failure code?

Reply:
Also, is this worth a log on failure?

Reply:
Good question. This was copied from the netebpfext code, but I see no reason for each extension to mask its error status. I've updated this module to return the error status as-is, and eBPF can apply its own policy to error statuses as necessary.

Since the error status is no longer masked, the existing log will capture the status.